src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticToken.java

```diff
@@ -0,0 +1,156 @@
+            SemanticTokenTypes.Event,
+            SemanticTokenTypes.Function,
+            SemanticTokenTypes.Method,
+            SemanticTokenTypes.Macro,
+            SemanticTokenTypes.Decorator,
+            // TODO: Need a mechanism for language server definitions to register token types and token modifiers
+            //  including how they should be interpreted for syntax highlighting and categorization
+            "member" // JavaScript/TypeScript-specific
+    );
+
+    // Semantic token modifiers that should be interpreted as representing identifier declarations
+    private static final Set<String> IDENTIFIER_DECLARATION_TOKEN_MODIFIERS = Set.of(
+            SemanticTokenModifiers.Declaration,
+            SemanticTokenModifiers.Definition
+    );
+
+    // Semantic token types that should be interpreted as representing keywords/reserved words in the language
+    private static final Set<String> KEYWORD_TOKEN_TYPES = Set.of(
+            SemanticTokenTypes.Keyword,
+            SemanticTokenTypes.Modifier
+    );
+
+    private final PsiFile file;
+    private final TextRange textRange;
+    private final String tokenType;
+    private final List<String> tokenModifiers;
+
+    private final LSPSemanticTokenElementType elementType;
+    private final LSPSemanticTokenPsiElement element;
+
+    /**
+     * Creates a new semantic token.
+     *
+     * @param file           the file
+     * @param textRange      the semantic token's text range
+     * @param tokenType      the token type
+     * @param tokenModifiers the token modifiers
+     */
+    LSPSemanticToken(
+            @NotNull PsiFile file,
+            @NotNull TextRange textRange,
+            @Nullable String tokenType,
+            @Nullable List<String> tokenModifiers
+    ) {
+        this.file = file;
+        this.textRange = textRange;
+        this.tokenType = tokenType;
+        this.tokenModifiers = tokenModifiers != null ? tokenModifiers : Collections.emptyList();
+
+        this.elementType = getElementType(this.tokenType, this.tokenModifiers);
+        this.element = new LSPSemanticTokenPsiElement(this);
+    }
+
+    @NotNull
+    PsiFile getFile() {
+        return file;
+    }
+
+    @NotNull
+    TextRange getTextRange() {
+        return textRange;
+    }
+
+    @Nullable
+    String getTokenType() {
+        return tokenType;
+    }
+
+    @NotNull
+    List<String> getTokenModifiers() {
+        return tokenModifiers;
+    }
+
+    @NotNull
+    LSPSemanticTokenElementType getElementType() {
+        return elementType;
+    }
+
+    @NotNull
+    private static LSPSemanticTokenElementType getElementType(@Nullable String tokenType,
+                                                              @NotNull List<String> tokenModifiers) {
+        if (tokenType != null) {
+            // If this is an identifier name token, see if it's a declaration or a reference
+            if (IDENTIFIER_NAME_TOKEN_TYPES.contains(tokenType)) {
```

Comment:
This is a big piece of the magic in this set of changes. We can use the combination of token type and token modifiers to understand how a specific portion of the file should be interpreted. The most important aspect is knowing that a given word is the name identifier for a declaration or a reference to a declaration, but we also know concretely that a given token is a comment, string or numeric literal, regular expression (which actually gives us a good opportunity for RegExp language injection), operator, etc.